sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java

```diff
@@ -235,4 +303,37 @@ public static ColumnarBatch toBatch(
     batch.setNumRows(n);
     return batch;
   }
+
+  /**
+   * <b>This method assumes that all constant column are at the end of schema
+   * and `constantColumnLength` represents the number of constant column.<b/>
+   *
+   * This method allocates columns to store elements of each field of the schema,
+   * the data columns use `OffHeapColumnVector` when `useOffHeap` is true and
+   * use `OnHeapColumnVector` when `useOffHeap` is false, the constant columns
+   * always use `ConstantColumnVector`.
+   *
+   * Capacity is the initial capacity of the vector, and it will grow as necessary.
+   * Capacity is in number of elements, not number of bytes.
+   */
+  public static ColumnVector[] allocateColumns(
```

Comment:
hmm I wonder why we need this method as there is only one call site for it in `VectorizedParquetRecordReader`.

Reply:
This method is required when making similar change to `Orc`. Need to move it into `VectorizedParquetRecordReader` in this pr ?

Reply:
@sunchao Sorry, When changing `OrcColumnarBatchReader`, I found that this method will only be used by `VectorizedParquetRecordReader`. I will submit a followup and move it to `VectorizedParquetRecordReader`